### PR TITLE
Scripts/Pet Implemented Argent Squire/Gruntling and Argent Pony Bridle

### LINKED ISSUE #4710: [Item] Argent Squire , Argent Pony Bridle
The item "Argent Pony Bridle" is bugged , when you click on it you should get option of:
- Bank
- Trader (vendor)
- Mailbox

But you can't see that list , 
You just see 
"Greetings name" 

In addtion there is item which call "Argent Pony Bridle" that should give mount for this companion what doesn't work as well

Related IDs:
http://old.wowhead.com/item=44998
http://old.wowhead.com/item=47541

Server info:
Hash: 090bb5a45e4b7608b57577c525a16416986a0b3e
Tdb: 335.11.43+

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/104381-item-argent-squire-argent-pony-bridle?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by sirikfoll**
**Changes proposed:**
- Based on Aokromes's script at #17783
-  Scripts/Pet Implemented Argent Squire/Gruntling and Argent Pony Bridle funcionalities(Shop/Mail/Bank) and the Pennants

**Target branch(es):** 3.3.5/6.x

**Issues addressed:** Closes #4710

**Tests performed:** Built and tested

### CODE REVIEW

**Comment by Aokromes** [comment]
simlt :) not me.

**Comment by tkrokli** [comment]
I love how you have handled the gossip, no hardcoded text like in #17783 :+1:

File: src/server/scripts/World/npcs_special.cpp [hunk 1]
```diff
@@ -2528,6 +2528,158 @@ class npc_train_wrecker : public CreatureScript
...
+                _check_timer = 0;
+            }
+            else
+                _check_timer -= diff;
+
+            if ((me->HasAura(SPELL_AURA_TIRED_S) || me->HasAura(SPELL_AURA_TIRED_G)) && me->HasFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_BANKER | UNIT_NPC_FLAG_MAILBOX | UNIT_NPC_FLAG_VENDOR))
+                me->RemoveFlag(UNIT_NPC_FLAGS, UNIT_NPC_FLAG_BANKER | UNIT_NPC_FLAG_MAILBOX | UNIT_NPC_FLAG_VENDOR);
+        }
+
+        bool IsArgentSquire() { return me->GetEntry() == NPC_ARGENT_SQUIRE; }
```
Comment (DDuarte) [inline src/server/scripts/World/npcs_special.cpp]: Minor but should be const

File: src/server/scripts/World/npcs_special.cpp [hunk 2]
```diff
@@ -2528,6 +2528,158 @@ class npc_train_wrecker : public CreatureScript
...
+                    if (!player->HasAura(SPELL_TIRED_PLAYER))
+                        player->CastSpell(player, SPELL_TIRED_PLAYER, true);
+                    break;
+                }
+                case GOSSIP_OPTION_DARNASSUS_SENJIN_PENNANT:
+                case GOSSIP_OPTION_EXODAR_UNDERCITY_PENNANT:
+                case GOSSIP_OPTION_GNOMEREGAN_ORGRIMMAR_PENNANT:
+                case GOSSIP_OPTION_IRONFORGE_SILVERMOON_PENNANT:
+                case GOSSIP_OPTION_STORMWIND_THUNDERBLUFF_PENNANT:
+                    if (me->GetEntry() == NPC_ARGENT_SQUIRE)
```
Comment (DDuarte) [inline src/server/scripts/World/npcs_special.cpp]: if (IsArgentSquire()) ?

**Comment by Nawuko** [comment]
I didnt looked into TaskScheduler yet and how they are implemented in TC. But are they thread safe? There are some crashports pointing to the lady deathwhisper PR which also uses TaskScheduler. And on the other site of thing how does task scheduler effect performance/memory usage. I really would be glad to get some opinions from @Shauren and @Naios.

And what is the current stand on Coding standards in regards to schedulers. If someone works on a PR should he/she consider using schedulers or eventmaps?

**Comment by Shauren** [comment]
I prefer to not use it. Unneccessary heap allocations, and the action code cannot be inlined

**Pull request merged**